### PR TITLE
ZkSaaS Gadget

### PR DESCRIPTION
**Pull request by tbraun96**

### CODE REVIEW

File: protocols/zk-saas/src/network/mod.rs [hunk 1]
```diff
@@ -244,10 +438,35 @@ fn handle_single_connection(
...
+        cert_der: Vec<u8>,
+    },
+    RegisterResponse {
+        id: RegistantId,
+        success: bool,
+        king_registry_id: RegistantId,
+    },
     // A message for the substrate gadget
-    SubstrateGadgetMessage { payload: GadgetProtocolMessage },
+    SubstrateGadgetMessage {
```
Comment (drewstone) [inline protocols/zk-saas/src/network/mod.rs:451]: What about this warrants a `Substrate` prefix? I want to be very meticulous with things like this. If we don't these prefixes (or maybe generalized ones can be better `BlockchainGadgetMessage`) I'm curious to hear more ideas.

**Review by drewstone — CHANGES_REQUESTED** [review]
LG curious your thoughts in certain places, @shekohex please review as well.

File: protocols/zk-saas/src/protocol/mod.rs [hunk 2]
```diff
@@ -0,0 +1,552 @@
...
+    ZkSaaSProofResult, ZkSaaSSystem,
+};
+use tangle_primitives::roles::RoleType;
+use tangle_primitives::verifier::to_field_elements;
+use tokio::sync::mpsc::UnboundedReceiver;
+
+pub mod proto_gen;
+
+type F = ark_bn254::Fr;
+type E = ark_bn254::Bn254;
```
[the remark this answers is not in the document]
Comment (shekohex) [reply protocols/zk-saas/src/protocol/mod.rs:42]: we will first need to add that information to the Runtime/on-chain types as a `curve` field [here](https://github.com/webb-tools/tangle/blob/9c6db763a6b5f9c0c81a8bbafeab98898928c4b6/primitives/src/jobs/zksaas.rs#L37) for example.

File: protocols/zk-saas/src/protocol/mod.rs [hunk 3]
```diff
@@ -0,0 +1,552 @@
...
+                    .ok_or_else(|| crate::Error::JobError {
+                        err: JobError {
+                            reason: "Phase one job not found".to_string(),
+                        },
+                    })?;
+
+                let JobType::ZkSaaSPhaseOne(phase_one) = phase_one_job.job_type else {
+                    return Err(Error::JobError {
+                        err: JobError {
+                            reason: "Phase one job type not ZkSaaS".to_string(),
```
[the remark this answers is not in the document]
Comment (shekohex) [reply protocols/zk-saas/src/protocol/mod.rs:102]: Yes, they are sanity checks.

File: protocols/zk-saas/src/protocol/mod.rs [hunk 4]
```diff
@@ -0,0 +1,552 @@
...
+                        },
+                    });
+                };
+
+                let JobType::ZkSaaSPhaseTwo(phase_two) = job.job_type else {
```
[the remark this answers is not in the document]
Comment (shekohex) [reply protocols/zk-saas/src/protocol/mod.rs:107]: They are not conditionins in per se, more like a refutable pattern matching using [let-else](https://doc.rust-lang.org/rust-by-example/flow_control/let_else.html) and ofc they all boils down to branching.

**Review by shekohex — CHANGES_REQUESTED** [review]
Looks good so far, waiting to see how the testing setup would look like

File: protocols/zk-saas/src/protocol/mod.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (drewstone) [reply protocols/zk-saas/src/protocol/mod.rs:103]: I think they can be removed, and if you think we need it I would suggest that we actually prevent this from happening on-chain. This feels more like an _insanity_ check TBH.

File: protocols/zk-saas/src/protocol/mod.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (drewstone) [reply protocols/zk-saas/src/protocol/mod.rs:108]: But I'm saying this is not needed. On line 81 we literally match to make sure its a ZkSaaSPhaseTwo
```
if matches!(job.job_type, JobType::ZkSaaSPhaseTwo(..)) {
```

[the remark this answers is not in the document]
Comment (drewstone) [reply protocols/zk-saas/src/protocol/mod.rs:108]: So this error will literally never be thrown

**Pull request merged**